al_log.js

```diff
@@ -191,47 +221,72 @@ function buildHostmeta(hostId, hostmetaElems, callback) {
  *      applicationId: 'o365'
  *  };
  *  Consult 'collected_message' definition in proto/common_proto.piqi.proto
+ * @param filterJson - JSON parsed object representing key-value pairs which 
+ * should exist in a message. For example, {type: 'Security', category: 'Admin'}.
+ * @param filterRegexp - regexp object filter. If content messages are JSON object
+ * then regexp filter if applied to a stringified message value.
  * @param callback
  * @returns
  */
 
-function buildMessages(content, parseContentFun, callback) {
-    async.reduce(content, [], function(memo, item, asyncCallback) {
-        var messageType = commonProtoPb.collected_message;
-        var messagePayload = parseContentFun(item);
-        buildType(messageType, messagePayload, function(err, buf) {
-            if (err) {
-                return asyncCallback(err);
+function buildMessages({content, parseContentFun, filterJson, filterRegexp}, callback) {
+    const jsonFilteredContent = alLogFilter.filterJson(content, filterJson);
+    const reFilter = alLogFilter.initRegExpFilter(filterRegexp);
+    
+    async.reduce(jsonFilteredContent, [], function(memo, item, asyncCallback) {
+        let messageType = commonProtoPb.collected_message;
+        let messagePayload = parseContentFun(item);
+        if (messagePayload && messagePayload.message) {
+            // Apply RegExp filter
+            const passFilter = reFilter ? reFilter.test(messagePayload.message) : true;
+            if (passFilter) {
+                const paramsMessage = {
+                    type: messageType,
+                    payload: messagePayload
+                };
+                buildType(paramsMessage, function(err, buf) {
+                    if (err) {
+                        return asyncCallback(err);
+                    } else {
+                        buf.messageLength = buf.message.length;
+                        memo.push(buf);
+                        return asyncCallback(err, memo);
```

Comment:
It should be  asyncCallback(null, memo);

Reply:
this like hasn't changed. And `err` here is `null`/`undefined`, see `if..else` condition